### PR TITLE
Expert filter properties

### PR DESCRIPTION
**Pull request by souissimai**

### CODE REVIEW

File: src/main/java/org/gridsuite/filter/expertfilter/expertrule/PropertiesExpertRule.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * @author maissa SOUISSI <maissa.souissi at rte-france.com>
+ */
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Setter
+@SuperBuilder
+public class PropertiesExpertRule extends AbstractExpertRule {
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/expertfilter/expertrule/PropertiesExpertRule.java:33]: In my opinion, PropertiesExpertRule can extend StringExpertRule, just adding propertyName

File: src/main/java/org/gridsuite/filter/expertfilter/expertrule/PropertiesExpertRule.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    public boolean evaluateRule(Identifiable<?> identifiable, FilterLoader filterLoader, Map<UUID, FilterEquipments> cachedUuidFilters) {
+        if (propertyName == null) {
+            return false;
+        }
+        String propertyValue = getFieldValue(this.getField(), propertyName, identifiable);
+        if (propertyValue == null) {
+            return false;
+        }
+        return switch (this.getOperator()) {
+            case EQUALS -> this.getPropertyValues().contains(propertyValue);
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/expertfilter/expertrule/PropertiesExpertRule.java:49]: equals and contains are not the same semantics

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java [hunk 3]
```diff
@@ -35,27 +35,27 @@ public final class ExpertFilterUtils {
 
     private ExpertFilterUtils() { }
 
-    public static <I extends Identifiable<I>> String getFieldValue(FieldType field, Identifiable<I> identifiable) {
+    public static <I extends Identifiable<I>> String getFieldValue(FieldType field, String propertyName, Identifiable<I> identifiable) {
         return switch (field) {
             case ID -> identifiable.getId();
             case NAME -> identifiable.getNameOrId();
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java:41]: can add 'case FREE_PROPERTIES -> identifiabe.getProperty(propertyName)' here

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java [hunk 4]
```diff
@@ -69,48 +69,55 @@ private static String getVoltageLevelFieldValue(FieldType field, VoltageLevel vo
                 VOLTAGE_LEVEL_ID_2 -> voltageLevel.getId();
             case LOW_VOLTAGE_LIMIT -> String.valueOf(voltageLevel.getLowVoltageLimit());
             case HIGH_VOLTAGE_LIMIT -> String.valueOf(voltageLevel.getHighVoltageLimit());
+            case FREE_PROPERTIES -> voltageLevel.getProperty(propertyName);
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java:72]: not needed as we add a case in getFieldValue

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java [hunk 5]
```diff
@@ -69,48 +69,55 @@ private static String getVoltageLevelFieldValue(FieldType field, VoltageLevel vo
...
                 VOLTAGE_LEVEL_ID_2,
-                NOMINAL_VOLTAGE_2 -> getVoltageLevelFieldValue(field, line.getTerminal(TwoSides.TWO).getVoltageLevel());
+                NOMINAL_VOLTAGE_2 -> getVoltageLevelFieldValue(field, null, line.getTerminal(TwoSides.TWO).getVoltageLevel());
             case SERIE_RESISTANCE -> String.valueOf(line.getR());
             case SERIE_REACTANCE -> String.valueOf(line.getX());
             case SHUNT_CONDUCTANCE_1 -> String.valueOf(line.getG1());
             case SHUNT_CONDUCTANCE_2 -> String.valueOf(line.getG2());
             case SHUNT_SUSCEPTANCE_1 -> String.valueOf(line.getB1());
             case SHUNT_SUSCEPTANCE_2 -> String.valueOf(line.getB2());
+            case FREE_PROPERTIES -> line.getProperty(propertyName);
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java:94]: not needed as we add a case in getFieldValue

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java [hunk 6]
```diff
@@ -69,48 +69,55 @@ private static String getVoltageLevelFieldValue(FieldType field, VoltageLevel vo
...
         return switch (field) {
             case COUNTRY,
                 NOMINAL_VOLTAGE,
-                VOLTAGE_LEVEL_ID -> getVoltageLevelFieldValue(field, load.getTerminal().getVoltageLevel());
+                VOLTAGE_LEVEL_ID -> getVoltageLevelFieldValue(field, null, load.getTerminal().getVoltageLevel());
             case P0 -> String.valueOf(load.getP0());
             case Q0 -> String.valueOf(load.getQ0());
             case CONNECTED -> getTerminalFieldValue(field, load.getTerminal());
             case LOAD_TYPE -> load.getLoadType().name();
+            case FREE_PROPERTIES -> load.getProperty(propertyName);
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java:110]: not needed as we add a case in getFieldValue

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java [hunk 7]
```diff
@@ -119,11 +126,13 @@ private static String getShuntCompensatorFieldValue(FieldType field, ShuntCompen
                 MAX_SUSCEPTANCE,
                 SWITCHED_ON_SUSCEPTANCE -> getSectionBasedFieldValue(field, shuntCompensator);
             case CONNECTED -> getTerminalFieldValue(field, shuntCompensator.getTerminal());
+            case FREE_PROPERTIES -> shuntCompensator.getProperty(propertyName);
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java:129]: not needed as we add a case in getFieldValue

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java [hunk 8]
```diff
@@ -140,8 +149,10 @@ private static String getGeneratorFieldValue(FieldType field, Generator generato
             case RATED_S -> String.valueOf(generator.getRatedS());
             case COUNTRY,
                 NOMINAL_VOLTAGE,
-                VOLTAGE_LEVEL_ID -> getVoltageLevelFieldValue(field, generator.getTerminal().getVoltageLevel());
+                VOLTAGE_LEVEL_ID -> getVoltageLevelFieldValue(field, null, generator.getTerminal().getVoltageLevel());
             case CONNECTED -> getTerminalFieldValue(field, generator.getTerminal());
+            case FREE_PROPERTIES -> generator.getProperty(propertyName);
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java:154]: not needed as we add a case in getFieldValue

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java [hunk 9]
```diff
@@ -204,25 +215,28 @@ private static String getSectionBasedFieldValue(FieldType field, ShuntCompensato
...
             case COUNTRY,
                     NOMINAL_VOLTAGE,
-                    VOLTAGE_LEVEL_ID -> getVoltageLevelFieldValue(field, battery.getTerminal().getVoltageLevel());
+                    VOLTAGE_LEVEL_ID -> getVoltageLevelFieldValue(field, null, battery.getTerminal().getVoltageLevel());
             case CONNECTED -> getTerminalFieldValue(field, battery.getTerminal());
             case MIN_P -> String.valueOf(battery.getMinP());
             case MAX_P -> String.valueOf(battery.getMaxP());
             case TARGET_P -> String.valueOf(battery.getTargetP());
             case TARGET_Q -> String.valueOf(battery.getTargetQ());
+            case FREE_PROPERTIES -> battery.getProperty(propertyName);
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java:228]: not needed as we add a case in getFieldValue

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java [hunk 10]
```diff
@@ -204,25 +215,28 @@ private static String getSectionBasedFieldValue(FieldType field, ShuntCompensato
...
             default -> throw new PowsyblException(FIELD_AND_TYPE_NOT_IMPLEMENTED + " [" + field + "," + battery.getType() + "]");
 
         };
     }
 
-    private static String getSubstationFieldValue(FieldType field, Substation substation) {
+    private static String getSubstationFieldValue(FieldType field, String propertyName, Substation substation) {
         return switch (field) {
             case COUNTRY -> String.valueOf(substation.getCountry().orElse(null));
+            case FREE_PROPERTIES -> substation.getProperty(propertyName);
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java:239]: not needed as we add a case in getFieldValue

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java [hunk 11]
```diff
@@ -278,6 +292,9 @@ private static String getTwoWindingsTransformerFieldValue(FieldType field, TwoWi
             case PHASE_REGULATING,
                 PHASE_REGULATION_MODE,
                 PHASE_REGULATION_VALUE -> getPhaseTapChangerFieldValue(field, twoWindingsTransformer.getPhaseTapChanger());
+            case FREE_PROPERTIES -> twoWindingsTransformer.getProperty(propertyName);
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java:295]: not needed as we add a case in getFieldValue

**Review by thangqp — APPROVED** [review]

File: src/main/java/org/gridsuite/filter/expertfilter/expertrule/PropertiesExpertRule.java (hunk 1 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/expertfilter/expertrule/PropertiesExpertRule.java:33]: done

File: src/main/java/org/gridsuite/filter/expertfilter/expertrule/PropertiesExpertRule.java (hunk 2 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/expertfilter/expertrule/PropertiesExpertRule.java]: done

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java (hunk 3 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java:41]: done

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java (hunk 4 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java]: done

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java (hunk 5 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java]: done

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java (hunk 7 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java]: done

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java (hunk 8 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java]: done

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java (hunk 9 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java]: done

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java (hunk 10 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java]: done

File: src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java (hunk 11 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/utils/expertfilter/ExpertFilterUtils.java]: done

**Review by ayolab — APPROVED** [review]

**Pull request merged**